client/src/reducers/teaReducers.js

```diff
@@ -1,20 +1,30 @@
-import { Set, Map } from "immutable";
-
-export default (state = [], action) => {
+export default (state = { allTeas: {}, teaIDs: [] }, action) => {
   switch (action.type) {
     case "ADD_TEA":
-      return [...state, action.payload];
-    case "DELETE_TEA":
-      return state.filter(t => t.id !== action.payload.id);
     case "EDIT_TEA":
-      return state.map(t => (t.id === action.payload.id ? action.payload : t));
+      let updateAllTeas = {
+        ...state.allTeas,
+        [action.payload.id]: action.payload
+      };
+      let updateIDsArr = [...state.teaIDs, action.payload.id];
+
+      return { allTeas: updateAllTeas, teaIDs: updateIDsArr };
+
+    case "DELETE_TEA":
+      let newIDsArr = state.teaIDs.filter(teaID => teaID !== action.payload);
+
+      return { ...state, teaIDs: newIDsArr };
+
     case "GET_TEAS":
-      // TODO: Data must be normalized.
-      // I wasn't planning on using immutable but it solved my problem here really well
-      // TODO figure out this TS error
-      return Set([...state, ...action.payload])
-        .map(Map)
-        .toJS();
+      let getAllTeas = {};
+      let getIDsArr = [];
+
+      action.payload.map(tea => {
+        getAllTeas[tea.id] = tea;
+        getIDsArr.push(tea.id);
+      });
+
+      return { allTeas: getAllTeas, teaIDs: getIDsArr };
```

Comment:
@JeffreyATW would it be appropriate here to check to see if the new teas in this payload differ from the existing teas in the store, and only return a new object when they differ?  Otherwise, won't we re-render every component that is connected to teas every time `GET_TEAS` is called, regardless of whether there have actually been any changes?

Reply:
Sure, that sounds like a good optimization. I generally blow away data when I receive a fresh copy from the server, but it wouldn't hurt to do a deep comparison to see if anything's actually changed. A better optimization would be to cut down on the amount of times `GET_TEAS` is called, if possible.